bundles/org.eclipse.osgi/container/src/org/eclipse/osgi/storage/BundleInfo.java

```diff
@@ -279,6 +280,105 @@ void setContent(File content, Type contentType) {
 			}
 		}
 
+		/**
+		 * Sets the content for this generation and checks for Eclipse-BundleShape header.
+		 * If the header is set to "dir", the content will be extracted to a directory.
+		 * This method creates the BundleFile to read headers without opening the JAR twice.
+		 * 
+		 * @param content the content file or directory
+		 * @param contentType the type of content
+		 * @return the BundleFile for this generation
+		 * @throws BundleException if an error occurs during installation
+		 */
+		BundleFile installBundleFile(File content, Type contentType) throws BundleException {
+			if (getBundleId() == 0) {
+				throw new UnsupportedOperationException("Should not be done for the system bundle."); //$NON-NLS-1$
+			}
+			synchronized (this.genMonitor) {
+				this.content = content;
+				this.isDirectory = content != null && Storage.secureAction.isDirectory(content);
+				this.contentType = contentType;
+				setLastModified(content);
+				return checkBundleShape();
+			}
+		}
+
+		/**
+		 * Creates the BundleFile and checks if the bundle should be extracted based on
+		 * the Eclipse-BundleShape header. If extraction is needed, swaps the content
+		 * to the extracted directory.
+		 * 
+		 * @return the BundleFile for this generation
+		 * @throws BundleException if an error occurs
+		 */
+		private BundleFile checkBundleShape() throws BundleException {
+			// Create the initial bundle file to read headers
+			BundleFile initialBundleFile;
+			if (getBundleId() == 0 && content == null && contentType != Type.CONNECT) {
+				initialBundleFile = new SystemBundleFile();
+			} else {
+				initialBundleFile = getStorage().createBundleFile(content, this, isDirectory, true);
+			}
+
+			// Only check for Eclipse-BundleShape if this is a DEFAULT type JAR file (not a directory)
+			if (contentType == Type.DEFAULT && content != null && content.isFile()) {
+				try {
+					initialBundleFile.open();
+					// Get headers to check for Eclipse-BundleShape
+					String bundleShape = getRawHeaders().get(Storage.ECLIPSE_BUNDLESHAPE);
+					if ("dir".equalsIgnoreCase(bundleShape)) { //$NON-NLS-1$
+						// Need to extract the JAR to a directory
+						initialBundleFile.close();
+						
+						File extractedDir = content.getParentFile();
+						File jarFile = content;
+						File tempExtractDir = new File(extractedDir, content.getName() + ".tmp"); //$NON-NLS-1$
+						
+						try {
+							// Extract to temporary directory first
+							getStorage().extractJarToDirectory(jarFile, tempExtractDir);
```

Comment:
Could use `initialBundleFile.getFile("", false)` to get the extracted content instead of having a new `extractJarToDirectory`. I'll admit the implementation of `getFile` when using the empty string root isn't the most efficient implementation and that could be improved.